signac/contrib/job.py

```diff
@@ -24,31 +23,173 @@
 logger = logging.getLogger(__name__)
 
 
-class _sp_save_hook:
-    """Hook to handle job migration when state points are changed.
-
-    When a job's state point is changed, in addition
-    to the contents of the file being modified this hook
-    calls :meth:`~Job._reset_sp` to rehash the state
-    point, compute a new job id, and move the folder.
-
-    Parameters
-    ----------
-    jobs : iterable of `Jobs`
-        List of jobs(instance of `Job`).
+class _StatepointDict(JSONDict):
+    """A JSON-backed dictionary for storing job statepoints.
 
+    There are three principal reasons for extending the base JSONDict:
+        1. Saving needs to trigger a job directory migration, and
+        2. Statepoints are assumed to not support external modification, so
+           they never need to load from disk _except_ the very first time a job
+           is opened by id and they're not present in the cache.
+        3. It must be possible to load and/or save on demand during tasks like
+           Job directory migrations.
```

Comment:
I actually kind of like capitalizing Job to indicate the class, but it's inconsistent with the rest of our docs.
```suggestion
           job directory migrations.
```